File: cards-agents-in-a-box/hearts/haskell/Game.hs
module Game where

import qualified Data.Vector as V
import Data.List
import Data.Maybe (fromJust) -- , isNothing, isJust, fromMaybe)

import System.Environment
import System.IO

clientRead = getLine
clientSendAndReceive = getLine

data Card = Card
  { suit :: String
  , rank :: String
  } deriving (Show)

data Trick = Trick
  { leader :: String
  , played :: [Card]
  } deriving (Show)

data Round = Round
  { number :: Int
  , tricks :: [Trick]
  , dealt :: [Card]
  , passed :: [Card]
  , received :: [Card]
  , held :: [Card]
  } deriving (Show)

data GameState = GameState
  { rounds :: [Round]
  } deriving (Show)

logit :: Show a => a -> IO ()
logit = print

currentTrick :: Round -> Trick
currentTrick = head . tricks

currentRound :: GameState -> Round
currentRound gs = head $ rounds gs

createRound :: GameState -> GameState
createRound gs =
  let newRound = Round{ number = length (rounds gs) + 1
                   , tricks = []
                   , dealt = []
                   , passed = []
                   , received = []
                   , held = [] }
  in gs{ rounds = newRound : rounds gs
     }

runTrick :: (GameState -> Card) -> GameState -> IO GameState
runTrick playCardFn gs = do
  logit "Starting trick"

  trick <- clientSendAndReceive "readyForTrick"

  let gsStartingTrick = gs{ rounds = (currentRound gs){ tricks = trick : tricks (currentRound gs) } : tail (rounds gs) }
  logit $ "Current trick:" ++ (show . currentTrick . currentRound) gsStartingTrick
  let cardToPlay = playCardFn gsStartingTrick
  logit $ "Playing card:" ++ (show cardToPlay)

  resultingTrick <- clientSendAndReceive "playCard" "card" cardToPlay
  logit $ "trick: result" ++ (show resultingTrick)
  let gsResultingTrick = gs{ rounds = (currentRound gs){ held = (delete cardToPlay (held (currentRound gs))), tricks = resultingTrick : tricks (currentRound gs) } : tail (rounds gs) }
  return gsResultingTrick

runRound :: (GameState -> [Card]) -> (GameState -> Card) -> GameState -> IO GameState
runRound passCardsFn playCardFn gs = do
  logit "Starting round"
  receivedCards <- clientSendAndReceive "readyForRound"
  let receivedCards' = V.toList receivedCards
      ags = gs{ rounds = (currentRound gs){ dealt = receivedCards', held = receivedCards' } : tail (rounds gs) }
  logit $ "You were dealt:" ++ (show . dealt . currentRound) ags
  newgs <- passCards passCardsFn ags
  playTrick playCardFn newgs

passCards :: (GameState -> [Card]) -> GameState -> IO GameState
passCards passCardsFn gs = do
  case number (currentRound gs) `rem` 4 of
    0 -> do
      logit "Not passing cards"
      return gs
    otherwise -> do
      logit "About to pass cards"
      let cardsToPass = passCardsFn gs
      received <- clientSendAndReceive "passCards" $ V.fromList cardsToPass
      let holding = (held (currentRound gs) \\ cardsToPass) ++ V.toList received
          ags = gs{ rounds = (currentRound gs){ held = holding, received = (V.toList received), passed = cardsToPass } : tail (rounds gs) }

      logit $ "Passing:" ++ show cardsToPass
      logit $ "Received cards:" ++ show received
      return ags

playTrick :: (GameState -> Card) -> GameState -> IO GameState
playTrick playCardFn gs = do
  loop 1 gs
  where
    loop :: Integer -> GameState -> IO GameState
    loop i gs | i <= 13 = do
                 newgs <- runTrick' gs
                 loop (i + 1) newgs
              | otherwise = return gs
    runTrick' = runTrick playCardFn

runGame :: (GameState -> [Card]) -> (GameState -> Card) -> IO ()
runGame passCardsFn playCardFn = do
  logit "Starting game"

  loop (GameState [])

  logit "done?"

  where
    loop :: GameState -> IO GameState
    loop gs = do
      newgs <- runRound passCardsFn playCardFn (createRound gs)
      roundResult <- clientSendAndReceive "finishedRound"
      logit $ "round result:" ++ show roundResult
      case roundResult of
        "nextRound" -> loop newgs
        "endGame" -> return newgs

play :: (GameState -> [Card]) -> (GameState -> Card) -> IO ()
play passCardsFn playCardFn = do
  hSetBuffering stdin NoBuffering
  logit "Starting"

  logit "playing"
  gameInfo <- clientRead
  logit $ "game info:" ++ show gameInfo

  runGame passCardsFn playCardFn

  logit "Game is over"
  gameResult <- clientSendAndReceive "finishedGame"
  logit $ "game result:" ++ show gameResult




--
-- finishGame handler ticket = do
--   print "Game result:"
--   gameResult <- get_game_result handler ticket
--   print gameResult
--   print "DONE"
--

File: cards-agents-in-a-box/skull/ruby/game.rb
require 'json'

def puts(*args)
  $stderr.puts *args
end

def p(*args)
  args.map!{|arg| arg.inspect}
  puts args
end

def print(*args)
  $stderr.print *args
end

class Client
  def send_command(command)
    $stdout << JSON.generate(command)
    $stdout << "\n"
    $stdout.flush
  end

  def read
    response = JSON.parse($stdin.gets.strip)
    if response["message"] == "error"
      puts "ERROR:", response
      exit
    end
    response
  end

  def send_and_receive(message, data={})
    send_command({message: message, data: data})
    read
  end
end

$client = Client.new

# class Card
#   attr_reader :suit, :rank
# 
#   def initialize(json)
#     @json = json
#     @suit = json["suit"]
#     @rank = json["rank"]
#   end
# 
#   def to_json(options={})
#     JSON.generate(@json)
#   end
# 
#   def to_s
#     "#{@rank} of #{@suit}"
#   end
# 
#   def ==(other)
#     @rank == other.rank && @suit == other.suit
#   end
#   alias_method :eql?, :==
# end
# 
# class Trick
#   attr_reader :number, :round, :leader, :played
#   def initialize(number, round, options)
#     @number = number
#     @round = round
#     @options = options
#     @leader = nil
#     @played = []
#   end
# 
#   def run
#     log "Starting trick"
#     trick = $client.send_and_receive("readyForTrick")["data"]
#     @leader = trick["leader"]
#     @played = trick["played"].map{|c| Card.new(c)}
# 
#     card_to_play = @options[:play_card_fn].(self)
#     @round.held.delete(card_to_play)
#     trick_result = $client.send_and_receive("playCard", card: card_to_play)["data"]
#     log "trick: result #{trick_result.inspect}"
#     @played = trick_result["played"].map{|c| Card.new(c)}
#   end
# 
#   def log(message)
#     @round.log "T:#{@number} #{message}"
#   end
# end

class Round
  attr_reader :number, :game #, :tricks, :dealt, :held
  def initialize(number, game, options)
    @number = number
    @game = game
    @options = options

    # @tricks = []
    # @dealt = []
    # @passed = []
    # @received = []
    # @held = []
  end

  # def create_trick
  #   trick_number = @tricks.size + 1
  #   trick = Trick.new trick_number, self, @options
  #   @tricks.push trick
  #   trick
  # end

  def run
    log "Starting round"

    $client.send_and_receive("readyForRound")
    log "You were dealt: #{hand.inspect}"
    @dealt = hand.dup.map{|c| Card.new(c)}
    @held = hand.dup.map{|c| Card.new(c)}

    pass_cards
    play_trick
  end

  # def pass_cards
  #   if @number % 4 == 0
  #     log "Not passing cards"
  #   else
  #     log "About to pass cards"
  #     cards_to_pass = @options[:pass_cards_fn].(self)

  #     @passed = cards_to_pass
  #     @held = @held.reject{|c| @passed.include?(c)}

  #     received_cards = $client.send_and_receive("passCards", cards: cards_to_pass)["data"]["cards"]
  #     log "Received cards: #{received_cards.inspect}"
  #     @received = received_cards.map{|c| Card.new(c)}
  #     @held = @held + @received
  #   end
  # end

  # def play_trick
  #   13.times do |trick_number|
  #     trick = create_trick
  #     trick.run
  #   end
  # end

  def log(message)
    @game.log "R:#{@number} #{message}"
  end
end

class Game
  def initialize(info, options)
    @info = info
    @options = options
    @rounds = []
  end

  def create_round
    round_number = @rounds.size + 1
    round = Round.new(round_number, self, @options)
    @rounds.push round
    round
  end

  def run
    log "Starting game"

    loop do
      round = create_round

      round.run

      round_result = $client.send_and_receive("finishedRound")["data"]
      log "round result: #{round_result.inspect}"
      break if round_result["status"] != "nextRound"
    end
  end

  def log(message)
    puts "P:#{@info.position} #{message}"
  end

  def self.do_play_card(&blk)
    @play_card_fn = blk
  end

  def self.do_pass_cards(&blk)
    @pass_cards_fn = blk
  end

  def self.play
    puts "playing"

    response = $client.read
    game_info = OpenStruct.new(response["data"])
    puts "game info: #{game_info.inspect}"

    game = Game.new game_info, {
      pass_cards_fn: @pass_cards_fn,
      play_card_fn: @play_card_fn
    }

    game.run
    puts "Game is over"
    game_result = $client.send_and_receive("finishedGame")["data"]
    puts "game result: #{game_result.inspect}"
  end

#   module Suit
#     SPADES = "spades"
#     HEARTS = "hearts"
#     DIAMONDS = "diamonds"
#     CLUBS = "clubs"
#   end
# 
#   module Rank
#     TWO = "two"
#     QUEEN = "queen"
#   end
end


